File: ruby/day4.rb
def main
  input = $stdin.read
  draws, *boards = input.split("\n\n")

  boards = boards.map { |raw| make_board(raw) }

  draws.split(",").each do |n|
    n = n.to_i
    if winner = call(boards, n)
      puts "part 1: #{score(winner, n)}"
      return
    end
  end
end

def score(board, n)
  unused = board[:tiles].reject { |n, tile| tile[:flipped] }.inject(0) { |sum, (n, _)| sum + n }
  unused * n
end

def call(boards, n)
  #puts "#{n}!"
  boards.each do |board|
    if tile = board[:tiles][n]
      tile[:flipped] = true
      return board if winner?(board)
    end
    #pretty_board(board)
  end
  nil
end

def pretty_board(board)
  board[:board].each do |row|
    puts row.map { |n| board[:tiles][n][:flipped] ? "X" : "O" }.join(" ")
  end
  puts "--"
end

def winner?(board)
  board[:board].each do |row|
    return true if row.all? { |n| board[:tiles][n][:flipped] }
  end
  (0..(board[:board][0].size-1)).each do |j|
    return true if board[:board].all? { |row| board[:tiles][row[j]][:flipped] }
  end
  false
end

def make_board(raw)
  tiles = {}
  board = []
  raw.lines.each_with_index do |line, i|
    row = []
    line.strip.split(/\s+/).each_with_index do |n, j|
      n = n.to_i
      row << n
      tiles[n] = {row: i, col: j}
    end
    board << row
  end
  {tiles: tiles, board: board}
end

main

File: src/day4.rs
use super::common::read_sections::read_sections;
use std::collections::HashMap;
use std::io::Read;

pub fn run<R: Read>(r: R) {
    let mut sections = read_sections(r);
    let order = sections.next().unwrap().remove(0);
    let mut boards = vec![];
    for section in sections {
        boards.push(parse_board(section));
    }
    let mut first_winner = None;
    let mut last_winner = None;
    for n in order.split(',') {
        let n = n.parse().unwrap();
        let (ww, ll): (Vec<Board>, Vec<Board>) = boards
            .into_iter()
            .map(|board| board.play(n))
            .partition(|board| board.is_winner());
        for w in ww.into_iter() {
            let score = n * w.unused();
            last_winner = Some(score);
            if first_winner.is_none() {
                first_winner = Some(score);
            }
        }
        boards = ll;
    }
    println!("part 1: {}", first_winner.unwrap());
    println!("part 2: {}", last_winner.unwrap());
}

struct Board {
    rows: [u8; 5],
    cols: [u8; 5],
    tiles: HashMap<u32, (usize, usize)>,
}

impl Board {
    fn play(mut self, n: u32) -> Self {
        match self.tiles.remove(&n) {
            None => self,
            Some((i, j)) => {
                let (mut rows, mut cols, tiles) = (self.rows, self.cols, self.tiles);
                rows[i] += 1;
                cols[j] += 1;
                Self { rows, cols, tiles }
            }
        }
    }

    fn is_winner(&self) -> bool {
        for r in self.rows {
            if r == 5 {
                return true;
            }
        }
        for c in self.cols {
            if c == 5 {
                return true;
            }
        }
        false
    }

    fn unused(&self) -> u32 {
        self.tiles.keys().sum()
    }
}

fn parse_board(raw: Vec<String>) -> Board {
    let mut tiles = HashMap::new();
    for (i, row) in raw.into_iter().enumerate() {
        for (j, n) in row.trim().split_whitespace().enumerate() {
            let n: u32 = n.parse().unwrap();
            tiles.insert(n, (i, j));
        }
    }
    Board {
        rows: [0; 5],
        cols: [0; 5],
        tiles,
    }
}
